### PR TITLE
perf tweaks

### PR DESCRIPTION
**Pull request by fkleuver**
@KFlash Here is some preliminary optimization work. In most cases it was just a matter of merging/reordering things in a more efficient matter. In some cases it seems a few validations could be skipped and I'm not sure if that's a safe thing to do. Mind doing a sanity check?

The benchmark shows a positive trend albeit not a very big one. But, the optimizations are pretty limited in scope so that's no surprise.

Bonus side effect it seems to make the library a bit smaller as the const enums aren't merged by default (not sure why though)

The benchmarks are a direct comparison between this branch (in my PR) and the master branch. So the only difference between the two versions are the changes in this PR

image
image
image

### CODE REVIEW

**Comment by KFlash** [comment]
This looks good, but maybe an idea to add a few comments so other devs better can understand what the code does?  I was thinking about optimizing like this before,  but reduced it to only tweak the leader code because not all have this knowledge. 

Beside this.  You good a great job with the optimization!!  Everything looks good :)

Regarding code size. You could merge all 4 function code into one mixed with some booleans. Example have a separate `parseFunctionAfterID` to reuse equal code instead of duplicates.  That way for method you just call that function.  Would need a few bitmask tweaks.  E. G.  Setting `InConstructor` if only in method context. 

For your worries.  If tests doesn't break.  Safe to remove.

**Comment by fkleuver** [comment]
Cool, thanks for taking a look.

File: src/parser/expression.ts [hunk 1]
```diff
@@ -2242,29 +2227,10 @@ function parseMethodDeclaration(state: ParserState, context: Context, objState:
...
 function parsePropertyMethod(state: ParserState, context: Context, objState: Modifiers): ESTree.FunctionExpression {
   // Create a new function scope
-  let functionScope = createScope(ScopeType.BlockStatement);
-
-  let id: ESTree.Identifier | null = null;
-  let firstRestricted: string | undefined;
+  const functionScope = createScope(ScopeType.BlockStatement);
   const { startIndex: start, startLine: line, startColumn: column } = state;
 
-  context =
```
Comment (KFlash) [inline src/parser/expression.ts]: @fkleuver I noticed a bug in Cherow 2.x that you could fix? Nothing wrong with your changes, it was me in the first place that seem to screw it up. The context masks you optimized should have been like this:

```js
 context =
    ((context | Context.InAwaitContext | Context.InYieldContext | Context.InArgList) ^
      (Context.InAwaitContext | Context.InYieldContext | Context.InArgList)) |
    (modifiers & ObjectFlags.Async ? Context.InAwaitContext : 0) |
    (modifiers & ObjectFlags.Generator ? Context.InYieldContext : 0) |
    Context.AllowNewTarget |
    Context.InMethod |
    Context.SuperProperty;
```
The `SuperProperty` isn't swapped, but only set. This allow this program to pass as valid  `(class A extends B { method() { super() } })`

You may change that

File: src/parser/declarations.ts [hunk 2]
```diff
@@ -224,23 +211,23 @@ export function parseHoistableFunctionDeclaration(
   let id: ESTree.Identifier | null = null;
   let name: string = '';
 
-  if (state.token & Token.IsIdentifier) {
+  if ((state.token & Token.IsIdentifier) === Token.IsIdentifier) {
```
Comment (KFlash) [inline src/parser/declarations.ts]: @fkleuver You could simply do `(state.token & 0x10FF) > 0x1000`, and then you also will make sure that both reserved and future reserved keywords are going through here as well.

**Comment by KFlash** [comment]
@fkleuver Merge this?

**Comment by KFlash** [comment]
@fkleuver Great work, but I think you forgot something? Look at [this line](https://github.com/cherow/cherow/blob/master/src/parser/expression.ts#L1760). And you can optimize / tweak it a little further too.

Something like this should do the trick, but... Maybe a comment or two would help? Most devs will fall off here, but performance is important!

```js
export function parseClassExpression(state: ParserState, context: Context): ESTree.ClassExpression {
  nextToken(state, context);
  // All class code is always strict mode implicitly
  context =
    (context = (context | 0b0000001000000000000_0100_00000000) ^ 0b0000001000000000000_0100_00000000) | Context.Strict;
  let id: ESTree.Expression | null = null;
  let superClass: ESTree.Expression | null = null;
  if (
    ((state.token & 0b0000000000000000001_0000_11111111) ^ 0b0000000000000000000_0000_01010100) >
    0b0000000000000000001_0000_00000000
  ) {
    validateIdentifier(state, context, BindingType.Class, state.token);
    id = parseIdentifier(state, context);
  }
  if ((state.token & 0b0000000000000000001_0000_11111111) > 0b0000000000000000001_0000_00000000) {
    optional(state, context | Context.AllowRegExp, Token.ExtendsKeyword);
    superClass = parseLeftHandSideExpression(state, context, false);
    context |= Context.SuperCall;
  } else {
    context = (context | Context.SuperCall) ^ Context.SuperCall;
  }

  const body = parseClassBodyAndElementList(state, context, BindingType.None, Origin.None);
  return {
    type: 'ClassExpression',
    id,
    superClass,
    body
  };
}
```

**Comment by fkleuver** [comment]
> but I think you forgot something?

I did not intend to do everything in one go, lol :)

> Maybe a comment or two would help? Most devs will fall off here

That's why I switched from hex to binary notation, so at least the values can be held directly against the const enum declaration and it's fairly easy to see what's what. I'd like to give them names but a few unions don't really have sensible names other than e.g. `IsYieldOrAsyncOrConstructorOrIdentifierOrReserved` kind of nonsense. I am expecting more answers to come as additional things are merged and perhaps start showing common patterns. Right now I lack the helicopter view to be able to go in a coherent manner, so kind of brute forcing it with limited scope now.

**Pull request merged**